config/config.go

```diff
@@ -499,9 +509,50 @@ func (cfg Config) ProducerPrivateKey() crypto.PrivateKey {
 			zap.Error(err),
 		)
 	}
+
+	if !cfg.whitelistSignatureScheme(sk) {
+		log.L().Panic("The private key's signature scheme is not whitelisted")
+	}
 	return sk
 }
 
+func (cfg Config) whitelistSignatureScheme(sk crypto.PrivateKey) bool {
+	var sigScheme string
+
+	switch sk.EcdsaPrivateKey().(type) {
+	case *ecdsa.PrivateKey:
+		sigScheme = SigP256k1
+	case *crypto.P256sm2PrvKey:
+		sigScheme = SigP256sm2
+	}
+
+	if sigScheme == "" {
+		return false
+	}
+	for _, e := range cfg.Chain.SignatureScheme {
+		if sigScheme == e {
+			// signature scheme is whitelisted
+			return true
+		}
+	}
+	return false
+}
+
+func (cfg Config) generateRandomKey() string {
```

Comment:
this function confused me. How could we randomly pick a scheme to generate a random key?

Reply:
this is to cover the corner case cfg.Chain.ProducerPrivKey is left empty, then generate a random key according to allowed schemes.

i think i'll leave it as empty, and not change the current code's behavior, please see the latest commit